SU2_CFD/src/interfaces/CInterface.cpp

```diff
@@ -56,240 +59,144 @@ CInterface::~CInterface(void) {
 
-        /*--- If the donor index is not -1 (this is, if the node is not originally a halo node) ---*/
-        if (Buffer_Recv_DonorIndices[iVertex] != -1) {
+    const auto nGlobalVertexDonor = displIdx.back() + nAllVertexDonor.back();
 
-          /*--- We set the donor index ---*/
-          Buffer_Bcast_Indices[iLocalVertex] = Buffer_Recv_DonorIndices[iVertex];
+    vector<unsigned long> donorIdx(nGlobalVertexDonor);
+    su2activematrix donorVar(nGlobalVertexDonor, nVar);
 
-          for (iVar = 0; iVar < nVar; iVar++)
-            Buffer_Bcast_Variables[iLocalVertex*nVar+iVar] = Buffer_Recv_DonorVariables[iVertex*nVar + iVar];
+    SU2_MPI::Allgatherv(sendDonorIdx.data(), sendDonorIdx.size(), MPI_UNSIGNED_LONG, donorIdx.data(),
+                        nAllVertexDonor.data(), displIdx.data(), MPI_UNSIGNED_LONG, MPI_COMM_WORLD);
 
+    SU2_MPI::Allgatherv(sendDonorVar.data(), sendDonorVar.size(), MPI_DOUBLE, donorVar.data(),
+                        nAllVarCounts.data(), displVar.data(), MPI_DOUBLE, MPI_COMM_WORLD);
 
-          iLocalVertex++;
+    /*--- This rank does not need to do more work. ---*/
+    if (markTarget < 0) continue;
 
-        }
-
-        if (iLocalVertex == TotalVertexDonor) break;
+    /*--- Sort the donor information by index to then use binary searches. ---*/
 
-      }
+    vector<size_t> order(donorIdx.size());
+    iota(order.begin(), order.end(), 0ul);
+    sort(order.begin(), order.end(), [&donorIdx](size_t i, size_t j) {return donorIdx[i] < donorIdx[j];} );
 
+    /*--- inplace permutation. ---*/
+    for (size_t i = 0; i < order.size(); ++i) {
+      auto j = order[i];
+      while (j < i) j = order[j];
+      if (i == j) continue;
+      swap(donorIdx[i], donorIdx[j]);
+      for (auto iVar = 0u; iVar < nVar; ++iVar)
+        swap(donorVar(i,iVar), donorVar(j,iVar));
     }
 
-    SU2_MPI::Bcast(Buffer_Bcast_Variables, nBuffer_BcastVariables, MPI_DOUBLE, MASTER_NODE, MPI_COMM_WORLD);
-    SU2_MPI::Bcast(Buffer_Bcast_Indices, nBuffer_BcastIndices, MPI_LONG, MASTER_NODE, MPI_COMM_WORLD);
-
-    long indexPoint_iVertex;
-    unsigned short iDonorPoint, nDonorPoints;
-    su2double donorCoeff;
-
-    /*--- For the target marker we are studying ---*/
-    if (Marker_Target >= 0) {
+    /*--- Loop over target vertices. ---*/
 
-      /*--- We have identified the local index of the Structural marker ---*/
-      /*--- We loop over all the vertices in that marker and in that particular processor ---*/
+    for (auto iVertex = 0ul; iVertex < target_geometry->GetnVertex(markTarget); iVertex++) {
+      const auto iPoint = target_geometry->vertex[markTarget][iVertex]->GetNode();
 
-      for (iVertex = 0; iVertex < target_geometry->GetnVertex(Marker_Target); iVertex++) {
+      if (!target_geometry->nodes->GetDomain(iPoint)) continue;
 
-        Point_Target = target_geometry->vertex[Marker_Target][iVertex]->GetNode();
+      auto& targetVertex = interpolator.targetVertices[markTarget][iVertex];
```

Comment:
Also just from looking at it, this looks way more appropriate than having it stored in the geometry.